effekt/shared/src/main/scala/effekt/core/Transformer.scala

```diff
@@ -169,25 +169,114 @@ object Transformer extends Phase[Typechecked, CoreTransformed] {
+
+    case source.New(impl) =>
       New(transform(impl, false))
 
     case source.Unbox(b) =>
       Unbox(transformAsPure(b))
 
+    case _ =>
+      transformUnbox(tree)
+  }
+  /**
+   * Transforms the source to a function block that expects to be called using [[core.Stmt.App]].
+   */
+  def transformAsControlBlock(tree: source.Term)(using Context): Block = tree match {
+    case v: source.Var =>
+      val sym = v.definition
+      val tpe = Context.blockTypeOf(sym)
+      tpe match {
+        case BlockType.FunctionType(tparams, cparams, vparamtps, bparamtps, restpe, effects) =>
+          // if this block argument expects to be called using PureApp or DirectApp, make sure it is
+          // by wrapping it in a BlockLit
+          lazy val targs = tparams.map(core.ValueType.Var.apply)
+          lazy val vparams: List[Param.ValueParam] = vparamtps.map{ t => Param.ValueParam(TmpValue(), transform(t))}
+          lazy val vargs = vparams.map{ case Param.ValueParam(id, tpe) => Pure.ValueVar(id, tpe) }
+          sym match {
+            case _: ValueSymbol => transformUnbox(tree)
+            case cns: Constructor =>
+              assert(bparamtps.isEmpty)
+              assert(effects.isEmpty)
+              assert(cparams.isEmpty)
+              BlockLit(tparams, Nil, vparams, Nil,
+                Stmt.Return(PureApp(BlockVar(cns), targs, vargs)))
+            case f: ExternFunction if isPure(f.capture) =>
+              assert(bparamtps.isEmpty)
+              assert(effects.isEmpty)
+              assert(cparams.isEmpty)
+              BlockLit(tparams, Nil, vparams, Nil,
+                Stmt.Return(PureApp(BlockVar(f), targs, vargs)))
+            case f: ExternFunction if pureOrIO(f.capture) =>
+              assert(effects.isEmpty)
+              val bparams: List[Param.BlockParam] = bparamtps.map{ t => Param.BlockParam(TmpBlock(), transform(t)) }
+              val bargs = bparams.map{
+                case Param.BlockParam(id, tpe) => Block.BlockVar(id, tpe, Set(id))
+              }
+              val result = TmpValue()
+              BlockLit(tparams, bparams.map(_.id), vparams, bparams,
+                core.Let(result, DirectApp(BlockVar(f), targs, vargs, bargs),
+                  Stmt.Return(Pure.ValueVar(result, transform(restpe)))))
+            case sym: BlockSymbol => BlockVar(sym)
+          }
+        case t: BlockType.InterfaceType =>
+          Context.abort(s"Expected a function but got an object of type ${t}")
+      }
+
     case source.BlockLiteral(tps, vps, bps, body) =>
       val tparams = tps.map(t => t.symbol)
       val cparams = bps.map { b => b.symbol.capture }
       BlockLit(tparams, cparams, vps map transform, bps map transform, transform(body))
 
+    case s @ source.Select(receiver, selector) =>
+      Member(transformAsObject(receiver), s.definition, transform(Context.inferredBlockTypeOf(tree)))
```

Comment:
I might be confusing things, but I hoped this could've been shared.

Reply:
It could, using yet another function like `transformAsSelectOrUnbox` (ideally with a better name). Since it's not too much/complicated I decided not to and just copied it, though I could just extract it.